flow/display_list_benchmarks.cc

```diff
@@ -0,0 +1,1000 @@
+      case SkPath::Verb::kLine_Verb:
+        GetLinesPath(path, sides, p, radius);
+        break;
+      case SkPath::Verb::kQuad_Verb:
+        GetQuadsPath(path, sides, p, radius);
+        break;
+      case SkPath::Verb::kConic_Verb:
+        GetConicsPath(path, sides, p, radius);
+        break;
+      case SkPath::Verb::kCubic_Verb:
+        GetCubicsPath(path, sides, p, radius);
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+std::string VerbToString(SkPath::Verb type) {
+  switch (type) {
+    case SkPath::Verb::kLine_Verb:
+      return "Lines";
+    case SkPath::Verb::kQuad_Verb:
+      return "Quads";
+    case SkPath::Verb::kConic_Verb:
+      return "Conics";
+    case SkPath::Verb::kCubic_Verb:
+      return "Cubics";
+    default:
+      return "Unknown";
+  }
+}
+
+// Draws a series of overlapping 20-sided polygons where the path segment
+// between each point is one of the verb types defined in SkPath.
+//
+// The number of polygons drawn will be varied to get an overall path
+// with approximately 20*N verbs, so we can get an idea of the fixed
+// cost of using drawPath as well as an idea of how the cost varies according
+// to the verb count.
+void BM_DrawPath(benchmark::State& state,
+                 std::unique_ptr<CanvasProvider> canvas_provider,
+                 SkPath::Verb type) {
+  DisplayListBuilder builder;
+  size_t length = kFixedCanvasSize;
+  canvas_provider->InitializeSurface(length, length);
+  auto canvas = canvas_provider->GetSurface()->getCanvas();
+
+  SkPath path;
+
+  std::string label = VerbToString(type);
+  SkPoint center = SkPoint::Make(length / 2.0f, length / 2.0f);
+  float radius = length * 0.25f;
+  state.SetComplexityN(state.range(0));
+
+  MultiplyPath(path, type, center, 20, state.range(0), radius);
+
+  state.counters["VerbCount"] = path.countVerbs();
+
+  builder.drawPath(path);
+  auto display_list = builder.Build();
+
+  // We only want to time the actual rasterization.
+  for (auto _ : state) {
+    display_list->RenderTo(canvas);
+    canvas_provider->GetSurface()->flushAndSubmit(true);
+  }
+
+  auto filename = canvas_provider->BackendName() + "-DrawPath-" + label + "-" +
+                  std::to_string(state.range(0)) + ".png";
+  canvas_provider->Snapshot(filename);
+}
+
+// Returns a set of vertices that describe a circle that has a
+// radius of `radius` and total vertex count of approximately
+// `vertex_count`. The final number of vertices may differ as
+// we always return an even number of vertices from this method.
+//
+// The resulting vertices will describe a disc consisting of a series
+// of triangles with two vertices on the circumference of the disc,
+// and the final vertex being the center point of the disc.
+//
+// Each vertex colour will alternate through Red, Green, Blue and Cyan.
+sk_sp<SkVertices> GetTestVertices(SkPoint center,
+                                  float radius,
+                                  size_t vertex_count,
```

Comment:
I think it's better to look at the vertex count because that's the data we actually input. We will need to figure out a way to either grab it, or approximate it. It looks like `SkVertices::approximateSize()` is a half-decent first-order approximation, as it returns an approximation for the total number of bytes used by the `SkVertices` object and all the data arrays. If we just subtract the size of the `SkVertices` object from that then we will have an approximation of the vertex count.

Reply:
We can also grab the vertex count from `Vertices.cc` in `dart:ui`: https://github.com/flutter/engine/blob/main/lib/ui/painting/vertices.cc#L61

It's a little annoying to have to plumb through that count ourselves but I think it's better than trying to calculate the number of triangles.

Reply:
We'll eventually get to that and flatten all of the Sk structures into the DL, but for now we rely on the data we can grab from the Sk objects.

Reply:
We can't grab the triangle count either, so I think it makes more sense to vary the vertex count to get an idea of that.